### ISSUE TITLE
Deleted Folders Appear in UI

### ISSUE THREAD

**Issue by chriswells0** [nextcloud/bookmarks#1820]
I have a semi-fresh install with no folders in the app (semi-fresh because I have had folders/bookmarks that are deleted now). I can add a "Test" folder, refresh, and see the folder sometimes but not others. Restarting Apache makes the refresh consistent. Then I can delete the folder, refresh, and see the folder sometimes but not others. Again, restarting Apache makes the refresh consistent.

I've tried:

* Holding Shift while refreshing to override the browser cache.
* Changing php.ini to set the following:
	* realpath_cache_ttl = 5
	* opcache.enable=0
	* opcache.enable_cli=0
* Waiting more than 15 minutes for all caches to expire.

Client:

 - OS: FreeBSD 12.3
 - Browser: Firefox 100.0

Server:

 - OS: FreeBSD 13.0
 - HTTP server: Apache 2.4.53
 - Database: MariaDB 10.5.15
 - PHP version: 8.0.18
 - Nextcloud version: 23.0.4
 - Bookmarks app version: 10.3.1

**Comment by chriswells0** [comment]
As an update: I received the notification that NC 24 was available for me the same day that I created this issue. I'm now running NC 24.0.1 and the same issue is occurring. Similarly, I can create a new folder, refresh repeatedly, and see the folder appear/disappear. It seems to be related to caching in some way.

**Comment by marcelklehr** [comment]
Hey @chriswells0 

I'm sorry for the trouble and will try to get to the bottom of this. Do you have more than one folder currently? So that sometimes all folders are displayed and sometimes no folders? Or is this indeed a case of only deleted folders still showing up somtimes? Can you access the JavaScript console and look for errors / related entries when deleting a folder, and later upon loading when the folder shows up out of nowhere? I will also try to replicate this as soon as possible (I'm travelling currently).

**Comment by chriswells0** [comment]
I have 0 folders currently. This was my initial setup, and I was experimenting to see how it worked and what would be a good strategy to sync both bookmarks and tabs. It would be nice to get advice on that, but no need to sidetrack here. I tried using root folders for both as well as having different base directories.

When I noticed it was behaving strangely (seeing a mix of those 2 approaches in the UI), I removed the accounts from floccus on my phone and laptop because I thought one must be syncing incorrectly in the background. With all accounts and even the app and extension deleted, I then removed all folders via the NC GUI. Only restarting Apache "resolved" the issue, but only until any change is made (e.g., creating or deleting a directory). I see consistent results aligned with the current state after Apache is restarted and until a change is made. I've also verified the directory is deleted from the DB while seeing it appear intermittently in the UI.

I have a near-perfect clone of this environment that I use to test NC upgrades as well as app installs/upgrades before making changes in prod. It has Bookmarks installed without any directories ever being created since I did not attempt to sync a device with the test environment. I haven't attempted to reproduce the issue there specifically because I thought you may want to try something in a very similar but more pristine environment. I can try any suggested tests/changes in either environment, though.

I'm off work this week, so I should have time to work on this. I'm also a developer, so I can perform any debugging you need. I simply don't know where to look for this app. I thought it might be useful to see what headers are being sent for the response that returns the folders, but I didn't see the folders in any of the XHR responses.

**Comment by marcelklehr** [comment]
>  I tried using root folders for both as well as having different base directories.

Different base directories is the way. Or at least a dedicated sub directory for tabs.

>  thought you may want to try something in a very similar but more pristine environment

I tried to reproduce this in a new clean environment and wasn't able to.

> I didn't see the folders in any of the XHR responses.

The folders are sent along [with the initial page load using Nextcloud's initialState API](https://github.com/nextcloud/bookmarks/blob/master/lib/Controller/WebViewController.php#L110). This may be where things go wrong, possibly. Are you caching HTML pages?

The folders are cached here: https://github.com/nextcloud/bookmarks/blob/3c22983f07908e6b4269c05b1b2c1f274d317b1a/lib/Db/TreeMapper.php#L709 using ACPu

It could also be that this functions fails to authenticate you: https://github.com/nextcloud/bookmarks/blob/master/lib/Controller/FoldersController.php#L467

**Comment by chriswells0** [comment]
The caching mechanism, APCu, was the clue I needed. I forgot about that setting in config.php. When I disabled it, the issue stopped happening. I found this on SO, and I was using mod_fcgid:

https://stackoverflow.com/questions/32270449/php-apcu-cgi-fastcgi-apc-fetch-always-return-false

More info:

https://stackoverflow.com/questions/34533695/is-the-new-apcu-apc-user-cache-shared-between-processes

After switching from mod_fcgid to mod_php (and from mod_mpm_worker to mod_mpm_prefork), I enabled APCu in NC again and the issue was resolved.

Thank you!
